### PR TITLE
py/obj: Add comments explaining the slot index scheme.

### PR DESCRIPTION
**Pull request by jimmo**
Based on suggestions from @stinos in https://github.com/micropython/micropython/pull/8813#issuecomment-1252440936

### CODE REVIEW

**Comment by jimmo** [comment]
@stinos I have also updated the wiki with some notes on using these macros. https://github.com/micropython/micropython/wiki/Build-Troubleshooting

I'm curious, what was the scenario where you needed to use `MP_OBJ_TYPE_SET_SLOT` (or was that just something you noticed). I was wondering whether I needed to add a `MP_OBJ_TYPE_UPDATE_SLOT` that doesn't require you to pass the index, i.e. it re-uses the existing index.

**Comment by stinos** [comment]
> updated the wiki 

looks good!

>  what was the scenario where you needed to use MP_OBJ_TYPE_SET_SLOT

I'm building types dynamically, sort of like in `mp_obj_new_type`, see [here](https://github.com/stinos/micropython-wrap/blob/master/classwrapper.h#L559).

> I was wondering whether I needed to add a MP_OBJ_TYPE_UPDATE_SLOT

Yeah I almost added it myself but then reasoned it might not see much usage. On the other hand, having to repeat the index is a pretty good recipe for bugs..

File: py/obj.h [hunk 1]
```diff
@@ -746,10 +746,16 @@ typedef struct _mp_obj_full_type_t {
...
+// slot_index_foo=0 to mean that the slot is unset. This also simplifies checking
+// if the slot is set. That means that we need to store index+1 in slot_index_foo
+// though and then access it as slots[slot_index_foo - 1]. This is an implementation
+// detail, the user of these macros doesn't need to be aware of it, and when using
+// MP_OBJ_TYPE_OFFSETOF_SLOT you should use zero-based indexing.
 #define MP_OBJ_TYPE_HAS_SLOT(t, f) ((t)->slot_index_##f)
 #define MP_OBJ_TYPE_GET_SLOT(t, f) (_MP_OBJ_TYPE_SLOT_TYPE_##f(t)->slots[(t)->slot_index_##f - 1])
 #define MP_OBJ_TYPE_GET_SLOT_OR_NULL(t, f) (_MP_OBJ_TYPE_SLOT_TYPE_##f(MP_OBJ_TYPE_HAS_SLOT(t, f) ? MP_OBJ_TYPE_GET_SLOT(t, f) : NULL))
-#define MP_OBJ_TYPE_SET_SLOT(t, f, v, n) ((t)->slot_index_##f = (n) + 1, (t)->slots[(t)->slot_index_##f - 1] = (void *)v)
+#define MP_OBJ_TYPE_SET_SLOT(t, f, v, n) ((t)->slot_index_##f = (n) + 1, (t)->slots[n] = (void *)v)
```
Comment (stinos) [inline py/obj.h:758]: Should be `slots[(n) + 1]` right?

Comment (stinos) [inline py/obj.h:758]: I don't think it matters functionally but for consistency use `slots[(n)]` ?

Comment (dpgeorge) [reply py/obj.h:758]: Yes, good to always wrap macro args in parenthesis.

Also, does this make any functional change?  The macro arg `n` is now evaluated twice.

Comment (jimmo) [reply py/obj.h:758]: It is always invoked with a literal.

Comment (dpgeorge) [reply py/obj.h:758]: OK.  But still worth wrapping `n` in parenthesis, if only for consistency with other macros.

Comment (jimmo) [reply py/obj.h]: Done (and rebased)

**Comment by dpgeorge** [comment]
Thanks, merged in d75c7e822cbfe7014fede1420f996a0af04bb13c

**Pull request closed without merging**